oauth_provider/tests/common_test_oauth_provider_controller.py

```diff
@@ -0,0 +1,420 @@
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue('Error: invalid_scope' in response.data)
+        self.assertTrue('An unknown error occured! Please contact your '
+                        'administrator' in response.data)
+
+    def test_authorize_error_wrong_uri(self):
+        """ Call the authorize method with a wrong redirect_uri
+
+        Must return an invalid_request error
+        """
+        self.login('demo', 'demo')
+        response = self.get_request('/oauth2/authorize', data={
+            'client_id': self.client.identifier,
+            'response_type': self.client.response_type,
+            'redirect_uri': 'http://wrong.redirect.uri',
+        })
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue('Error: invalid_request' in response.data)
+        self.assertTrue('Mismatching redirect URI' in response.data)
+
+    def test_authorize_error_missing_uri(self):
+        """ Call /oauth2/authorize without any configured redirect URI
+
+        Must return an invalid_request error
+        """
+        self.client.redirect_uri_ids.unlink()
+        self.login('demo', 'demo')
+        response = self.get_request('/oauth2/authorize', data={
+            'client_id': self.client.identifier,
+            'response_type': self.client.response_type,
+            'scope': self.client.scope_ids[0].code,
+        })
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue('Error: invalid_request' in response.data)
+        self.assertTrue('Missing redirect URI.' in response.data)
+
+    def test_authorize_post_errors(self):
+        """ Call /oauth2/authorize in POST without any session
+
+        Must return an unknown client identifier error
+        """
+        self.login('demo', 'demo')
+        response = self.post_request('/oauth2/authorize')
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue('Unknown Client Identifier!' in response.data)
+        self.assertTrue('This client identifier is invalid.' in response.data)
+
+    @mock.patch('odoo.http.WebRequest.env', new_callable=mock.PropertyMock)
+    def test_authorize_unsafe_chars(self, request_env):
+        """ Call /oauth2/authorize with unsafe chars in the query string """
+        # Mock the http request's environ to allow it to see test records
+        request_env.return_value = self.env(user=self.user)
+
+        query_string = 'client_id=%s&response_type=%s&state={}' % (
+            self.client.identifier,
+            self.client.response_type,
+        )
+        self.login('demo', 'demo')
+        response = self.test_client.get(
+            '/oauth2/authorize', query_string=query_string,
+            environ_base=self.werkzeug_environ)
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue(self.client.name in response.data)
+
+
+class TestOauthProviderRefreshTokenController(object):
+    def test_refresh_token_error_too_much_scopes(self):
```

Comment:
@sylvain-garancher - this one is failing, and I'm actually kind of confused on the purpose of it. What is invalid about this request? These are valid scopes in the client, and multiple scopes should be supported at once. 

I feel like I'm missing something, particularly because there is a nearly identical request below in `test_refresh_token`, except that the codes are joined by a  ` `